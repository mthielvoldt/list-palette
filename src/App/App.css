

.container {
  width: 100%;
  height: auto;
  min-height: 500px;
  max-width: 500px;
  min-width: 450px;
  background: #e4e4e4;
  border-radius: 20px;
  box-shadow: 4px 3px 7px 2px #00000040;
  padding: 1rem;
  box-sizing: border-box;
}

.heading {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 1rem;
}
.heading h1 {
  transform: rotate(2deg);
  padding: 0.2rem 1.2rem;
  border-radius: 20% 5% 20% 5%/5% 20% 25% 20%;
  background-color: #fcf876;
  font-size: 1.5rem;
}