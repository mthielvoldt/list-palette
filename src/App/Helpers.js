
function dupHelper(id, location, items) {

    // Start with copying the root item - 
    // putting the copy at the end of the previous array
    let newParentItem = { ...items[id], id: items.length};
    if (items[id].child != null) {
        newParentItem.child = items.length+1; 
    }

    // shallow-copy items array, inserting the duplicated root. 
    items.push(newParentItem);

    // connectItem expects to be able to mutate the array iteself
    // but not the linked-to objects in the array.  It makes new copies of those if they will be changed. 
    connectItem(newParentItem.id, location, items, "child");

    /**Now, we recursively copy all the children of the parent item, 
     * starting with the parent item's first child. 
     * origId = items[id].child (the original to copy = first child of duplicated root)
     * newId = items.length (the end of the new array; the index for new elements)
     * parent = newItem.id (the new root).
     * previous = null - item will be located at the top of its list. 
     */
    if (items[id].child != null) {
        copyItem(items[id].child, items.length, newParentItem.id, null);
    }

    function copyItem(origId, newId, parent, previous) {
        let origItem = items[origId];
        let newItem = { ...origItem, id: newId, parent: parent, previous: previous, child: null, next: null }
        items.push(newItem);

        if (origItem.child != null) {
            newItem.child = newId + 1;
            newId = copyItem(origItem.child, newId + 1, newId, null);
        }
        if (origItem.next != null) {
            //console.log("newItem:", newItem, "newItemCopy:", newItemCopy);
            newItem.next = newId + 1; // at this point, newId will have been incremented by every child above. 
            newId = copyItem(origItem.next, newId + 1, parent, newItem.id);
        }
        //console.log("origItem, newItem:", origItem, newItem);
        return newId;
    }
}

function mergeHelper(src, dest, items) {
    let i = items[src].child;
    while (i != null) {
        let j = items[dest].child;
        while (j != null) {
            if (items[j].text === items[i].text) {
                mergeHelper(i, j, items);
                break;
            }
            j = items[j].next;
        }
        if (j == null) {   // no match
            dupHelper(i, dest, items);
        }
        i = items[i].next;
    }
}

function connectItem(srcId, destId, items, relation) {
    // This function replaces 3 elements of the 'items' array with new objects: 
    // items[srcId], items[destId], and a third item that depends on the relation parameter. 
    // The original objects referenced by the 'items' array are not mutated, but 
    // the 'items' array itself (the object references) is mutated. 

    // make copy of src and dest. 
    let srcItem = { ...items[srcId] };
    let destItem = { ...items[destId] };

    switch (relation) {

      case "before":  // plug in srcItem immediately before destItem
        srcItem.previous = destItem.previous;
        srcItem.parent = destItem.parent;
        srcItem.next = destItem.id;

        // handle connection above source item
        if (destItem.previous === null) {
          items[destItem.parent] = { ...items[destItem.parent], child: srcItem.id };
        } else {  // dest was not first item
          items[destItem.previous] = { ...items[destItem.previous], next: srcItem.id };
        }
        destItem.previous = srcItem.id;
        break;

      case "after": // plug in srcItem immediately after destItem
        srcItem.previous = destItem.id;
        srcItem.parent = destItem.parent;
        srcItem.next = destItem.next;

        if (destItem.next != null) {
          items[destItem.next] = { ...items[destItem.next], previous: srcItem.id };
        }
        destItem.next = srcItem.id;
        break;

      case "child": // plug in srcItem as first child of destItem
        srcItem.previous = null;
        srcItem.parent = destItem.id;
        srcItem.next = destItem.child;

        if (destItem.child != null) {
          items[destItem.child] = { ...items[destItem.child], previous: srcItem.id };
        }
        destItem.child = srcItem.id;
        break;

      default:
        console.error("unspefified relationship to destination");
        return -1;
    }

    // plug srcItem and destItem with updated links back into array. 
    items[srcItem.id] = srcItem;
    items[destItem.id] = destItem;
  }

  // Take the identified item out of the flow by patching the links to cut it out. 
  function disconnectItem(id, items) {
    // This function mutates the 'items' array by replacing the object references, 
    // but it does not mutate any of the objects pointed to by the 'items' elements. 

    let item = items[id]; // syntactic convenience.  Item is not mutated; 

    // is the item the first in its list? 
    // then we have to make its parent point to it's next sibling
    if (item.previous === null) {
      items[item.parent] = { ...items[item.parent], child: item.next };
    } else {
      items[item.previous] = { ...items[item.previous], next: item.next };
    }

    // Patch reverse link
    if (item.next !== null) {
      items[item.next] = { ...items[item.next], previous: item.previous };
    }
  }

export {
    dupHelper,
    mergeHelper,
    connectItem,
    disconnectItem,
}