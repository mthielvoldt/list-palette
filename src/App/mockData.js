
const mockData = {};

mockData.itemsShort = [
  { "id": 3, "text": "People", "child": null, "next": 2, "checked": "false" },
  { "id": 2, "text": "To-do", "child": null, "next": 1, "checked": "false" },
  { "id": 1, "text": "Groceries", "child": null, "next": null, "checked": "false" },
  { "id": 0, "text": "Home", "child": 3, "next": null, "checked": "false" }
];

mockData.nestedResponse = {
  "status": 200,
  "data": {
    "user": "Daryl",
    "items": [
      { "id": 0, "text": "Home", "child": 3, "next": null, "checked": "false" },
      { "id": 6, "text": "alpha.charlie", "child": null, "next": null, "checked": "false" },
      { "id": 5, "text": "alpha.beta", "child": null, "next": 6, "checked": "false" },
      { "id": 4, "text": "alpha.alpha", "child": null, "next": 5, "checked": "false" },
      { "id": 3, "text": "alpha", "child": 4, "next": 2, "checked": "unchecked" },
      { "id": 7, "text": "beta.charlie", "child": null, "next": null, "checked": "false" },
      { "id": 8, "text": "beta.beta", "child": null, "next": 7, "checked": "false" },
      { "id": 9, "text": "beta.alpha", "child": null, "next": 8, "checked": "false" },
      { "id": 2, "text": "beta", "child": 9, "next": 1, "checked": "false" },
      { "id": 10, "text": "charlie.foxtrot", "child": null, "next": null, "checked": "false" },
      { "id": 11, "text": "charlie.echo", "child": null, "next": 10, "checked": "false" },
      { "id": 12, "text": "charlie.delta", "child": null, "next": 11, "checked": "false" },
      { "id": 13, "text": "charlie.charlie", "child": null, "next": 12, "checked": "false" },
      { "id": 14, "text": "charlie.beta", "child": null, "next": 13, "checked": "false" },
      { "id": 15, "text": "charlie.alpha", "child": null, "next": 14, "checked": "false" },
      { "id": 1, "text": "charlie", "child": 15, "next": null, "checked": "unchecked" }
    ]
  },
};

export default mockData;