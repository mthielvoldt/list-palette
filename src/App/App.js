import React, { useState } from "react";
import AddItemForm from "../AddItemForm/AddItemForm";
import List from "../List/List";
import mockQuery from '../data.js'
import './App.css';

let nextId = mockQuery.reduce( (accum, current) => 
  ((current.id > accum) ? current.id : accum), 0);
console.log("nextID:", nextId);

let initialState = [];
mockQuery.forEach((elem, index) => {initialState[elem.id] = elem});
console.log(initialState);

function App() {
  const [items, setItems] = useState(initialState);

  function addItem(newText) {
    let newItem = { id: ++nextId, text: newText, children: [] };
    setItems(prevItems => [...prevItems, newItem]);
  }

  function deleteItem(id) {
    setItems(prevItems => prevItems.filter(item => item.id !== id));
  }

  function moveItem(source, dest) {
    console.log("reOrder: src: %d dest: %d", source, dest);
    const newItems = [...items];
    //newItems.splice(source, 1); // remove source from array;
    newItems.splice(dest, 0, items[source]);
    console.log(newItems);

    setItems(newItems);
  }

  return (
    <div className="container">
      <div className="heading">
        <h1>To-Do List</h1>
      </div>
      <AddItemForm callback={addItem} />
      <List items={items} index={items[0].child} deleteCB={deleteItem} moveCB={moveItem} />
    </div>
  );
}

export default App;

