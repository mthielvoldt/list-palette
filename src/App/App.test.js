import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import testUtils from 'react-dom/test-utils';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import App from './App';
import mockData from './mockData'
import axios from 'axios';
jest.mock('axios');

expect.extend({ toMatchDiffSnapshot });

afterEach(cleanup);



describe("The App ", () => {

  it("renders with anonymous user", async () => {
    axios.get.mockResolvedValue({ data: { user: null, items: mockData.itemsShort } });
    //axios.get.mockImplementation(() => Promise.resolve(resp))

    const app = render(<App />);
    const peopleItem = await app.findByText(/People/i);
    expect(peopleItem).toBeInTheDocument();
    expect(app.getByText(/Home>/i)).toBeInTheDocument();
    expect(app.getByText(/Sign in/i)).toBeInTheDocument();
    expect(app.asFragment()).toMatchSnapshot();
    expect(axios.get).toBeCalledWith('/items');
  });

  it("renders with signed-in user", async () => {
    axios.get.mockResolvedValue({ data: { user: 'John', items: mockData.itemsShort } });
    const app = render(<App />);
    //axios.get.mockImplementation(() => Promise.resolve(resp))


    const userLink = await app.findByText(/John/i);
    expect(userLink).toBeInTheDocument();
    expect(app.queryByText(/Sign in/i)).toBe(null);
    expect(app.asFragment()).toMatchSnapshot();
  });

  it("renders and sends PUT when user adds items", async () => {
    // mock the PUT request that should be sent when item is added.
    axios.put.mockResolvedValue({ data: {}, status: 200 });

    // Render and wait for re-load of data (this is asynch)
    const app = render(<App />);
    await app.findByText('Groceries');

    // Simulate user typing something into input
    const addItemInput = app.getByTestId('add-item-input');
    fireEvent.change(addItemInput, { target: { value: "Silly String" } });

    // We haven't submitted yet, so make sure we don't see the new string yet. 
    var newItem = app.queryByText('Silly String');
    expect(newItem).toBe(null);

    // Now we click the button. 
    fireEvent.click(app.getByText("Add"));

    // Does the new item render?
    newItem = app.queryByText('Silly String');
    expect(newItem).toBeVisible();

    // Does the API get hit with the correct payload? 
    expect(axios.put).lastCalledWith('/items',
      {
        "insert": [
          { "id": 4, "text": "Silly String", "checked": false, "child": null, "previous": null, "parent": 0, "next": 3 }],
        "update": [
          { "id": 0, "text": "Home", "checked": "false", "next": null, "child": 4, "parent": null, "previous": null }]
      });

    // If we just click the button again, another PUT request should not happen. 
    fireEvent.click(app.getByText("Add"));

    // If we change the text then press enter, we should see a new API hit. 
    fireEvent.change(addItemInput, { target: { value: "Crazy Cheeze" } });
    fireEvent.keyPress(addItemInput, { key: 'Enter', charCode: 13 });

    newItem = app.queryByText("Crazy Cheeze");
    expect(newItem).toBeVisible();

    expect(axios.put).lastCalledWith('/items',
      {
        "insert": [
          { "id": 5, "text": "Crazy Cheeze", "checked": false, "child": null, "previous": null, "parent": 0, "next": 4 }],
        "update": [
          { "id": 0, "text": "Home", "checked": "false", "next": null, "child": 5, "parent": null, "previous": null }]
      });

    expect(axios.put).toBeCalledTimes(2);
  });


  it("renders and sends PUT when user moves items", async () => {
    // mock the GET request that loads the seed data.
    axios.get.mockResolvedValue(mockData.nestedResponse);
    // mock the PUT request that should be sent when item is moved.
    axios.put.mockResolvedValue({ data: {}, status: 200 });

    // Mock the dataTransfer object (because it isn't implemented in js-DOM). 
    // Create a local object to hold key-value pairs, and functions for writing and retreiving. 
    let storedData = {};
    const mockSetData = jest.fn((key, value) => {
      Object.assign(storedData, { [key]: value })
    })
    const mockGetData = jest.fn(key => storedData[key]);

    // Render and wait for loading the seed items
    const app = render(<App />);
    let alpha = await app.findByText('alpha');
    let beta = app.getByText('beta');
    let appBefore = app.asFragment();

    // Simulate user moving an item on first level 
    testUtils.Simulate.dragStart(alpha, { dataTransfer: { setData: mockSetData } });
    testUtils.Simulate.dragOver(beta, { dataTransfer: { getData: mockGetData } });
    testUtils.Simulate.drop(beta, { dataTransfer: { getData: mockGetData } });

    expect(app.asFragment()).toMatchDiffSnapshot(appBefore);
    expect(axios.put).lastCalledWith('/items',
      {
        "insert": [],
        "update": [
          { "id": 0, "text": "Home", "child": 2, "next": null, "checked": "false", "parent": null, "previous": null },
          { "id": 3, "text": "alpha", "child": 4, "next": 1, "checked": "unchecked", "parent": 0, "previous": 2 },
          { "id": 2, "text": "beta", "child": 9, "next": 3, "checked": "false", "parent": 0, "previous": null }
        ]
      });

    // If we just drop the item back onto itself, the API should not be hit. 
  });

  it.skip("moves items with touch only after long-touch", async () => {
    // touch item and move immediately
    
    // it should not be "lifted" (on a phone it would scroll)

    // touch and hold item. 
    // it should lift item for drag/drop
  })

  it.skip("renders and sends PUT when user deletes items", async () => {
  });

  it.skip("renders and sends PUT when user merges items", async () => {
  });

  it.skip("handles a sign-up", async () => {
  });



});



