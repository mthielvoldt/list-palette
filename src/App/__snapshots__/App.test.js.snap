// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App  renders and sends PUT when user moves items 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -107,11 +107,11 @@
          >
            <path
              d=\\"M12.14 8.753l-5.482 4.796c-.646.566-1.658.106-1.658-.753V3.204a1 1 0 0 1 1.659-.753l5.48 4.796a1 1 0 0 1 0 1.506z\\"
            />
          </svg>
-         beta
+         alpha
          <img
            alt=\\"delete icon\\"
            class=\\"icon\\"
            src=\\"/icons/delete-icon.svg\\"
          />
@@ -145,11 +145,11 @@
          >
            <path
              d=\\"M12.14 8.753l-5.482 4.796c-.646.566-1.658.106-1.658-.753V3.204a1 1 0 0 1 1.659-.753l5.48 4.796a1 1 0 0 1 0 1.506z\\"
            />
          </svg>
-         alpha
+         beta
          <img
            alt=\\"delete icon\\"
            class=\\"icon\\"
            src=\\"/icons/delete-icon.svg\\"
          />"
`;

exports[`The App  renders with anonymous user 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="navbar navbar-expand-md navbar-dark bg-dark"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        <img
          alt="List palette logo"
          class="mr-2"
          height="48"
          src="logo192.svg"
        />
        List Palette
      </a>
      <button
        aria-controls="navbarCollapse"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-target="#navbarCollapse"
        data-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarCollapse"
      >
        <ul
          class="navbar-nav mr-auto"
        >
          <li
            class="nav-item"
          >
            <button
              class="nav-link"
            >
              About
            </button>
          </li>
        </ul>
        <button
          class="btn btn-success mx-1 my-2 my-sm-0"
        >
          Sign in
        </button>
        <button
          class="btn btn-outline-success my-2 my-sm-0"
        >
          Sign up
        </button>
      </div>
    </nav>
  </header>
  <div
    class="container"
  >
    <div
      class="location"
    >
      <button
        class="location-button"
      >
        Home&gt;
      </button>
    </div>
    <div
      class="form"
    >
      <input
        data-testid="add-item-input"
        type="text"
        value=""
      />
      <button>
        <span
          class="add-span"
        >
          Add
        </span>
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        draggable="true"
      >
        <svg
          class="bi bi-caret-right-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        People
        <img
          alt="delete icon"
          class="icon"
          src="/icons/delete-icon.svg"
        />
        <img
          alt="edit icon"
          class="icon"
          src="/icons/edit-icon.svg"
        />
        <img
          alt="focus icon"
          class="icon"
          src="/icons/focus-icon.svg"
        />
        <img
          alt="merge icon"
          class="icon"
          src="/icons/merge-icon.svg"
        />
      </div>
      <div
        class="item"
        draggable="true"
      >
        <svg
          class="bi bi-caret-right-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        To-do
        <img
          alt="delete icon"
          class="icon"
          src="/icons/delete-icon.svg"
        />
        <img
          alt="edit icon"
          class="icon"
          src="/icons/edit-icon.svg"
        />
        <img
          alt="focus icon"
          class="icon"
          src="/icons/focus-icon.svg"
        />
        <img
          alt="merge icon"
          class="icon"
          src="/icons/merge-icon.svg"
        />
      </div>
      <div
        class="item"
        draggable="true"
      >
        <svg
          class="bi bi-caret-right-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        Groceries
        <img
          alt="delete icon"
          class="icon"
          src="/icons/delete-icon.svg"
        />
        <img
          alt="edit icon"
          class="icon"
          src="/icons/edit-icon.svg"
        />
        <img
          alt="focus icon"
          class="icon"
          src="/icons/focus-icon.svg"
        />
        <img
          alt="merge icon"
          class="icon"
          src="/icons/merge-icon.svg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`The App  renders with signed-in user 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="navbar navbar-expand-md navbar-dark bg-dark"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        <img
          alt="List palette logo"
          class="mr-2"
          height="48"
          src="logo192.svg"
        />
        List Palette
      </a>
      <button
        aria-controls="navbarCollapse"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-target="#navbarCollapse"
        data-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarCollapse"
      >
        <ul
          class="navbar-nav mr-auto"
        >
          <li
            class="nav-item"
          >
            <button
              class="nav-link"
            >
              About
            </button>
          </li>
        </ul>
        <div
          class="navbar-nav"
        >
          <button
            class="nav-link mr-1"
          >
            John
          </button>
          <button
            class="btn btn-secondary my-2 my-sm-0"
          >
            Sign out
          </button>
        </div>
      </div>
    </nav>
  </header>
  <div
    class="container"
  >
    <div
      class="location"
    >
      <button
        class="location-button"
      >
        Home&gt;
      </button>
    </div>
    <div
      class="form"
    >
      <input
        data-testid="add-item-input"
        type="text"
        value=""
      />
      <button>
        <span
          class="add-span"
        >
          Add
        </span>
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        draggable="true"
      >
        <svg
          class="bi bi-caret-right-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        People
        <img
          alt="delete icon"
          class="icon"
          src="/icons/delete-icon.svg"
        />
        <img
          alt="edit icon"
          class="icon"
          src="/icons/edit-icon.svg"
        />
        <img
          alt="focus icon"
          class="icon"
          src="/icons/focus-icon.svg"
        />
        <img
          alt="merge icon"
          class="icon"
          src="/icons/merge-icon.svg"
        />
      </div>
      <div
        class="item"
        draggable="true"
      >
        <svg
          class="bi bi-caret-right-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        To-do
        <img
          alt="delete icon"
          class="icon"
          src="/icons/delete-icon.svg"
        />
        <img
          alt="edit icon"
          class="icon"
          src="/icons/edit-icon.svg"
        />
        <img
          alt="focus icon"
          class="icon"
          src="/icons/focus-icon.svg"
        />
        <img
          alt="merge icon"
          class="icon"
          src="/icons/merge-icon.svg"
        />
      </div>
      <div
        class="item"
        draggable="true"
      >
        <svg
          class="bi bi-caret-right-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        Groceries
        <img
          alt="delete icon"
          class="icon"
          src="/icons/delete-icon.svg"
        />
        <img
          alt="edit icon"
          class="icon"
          src="/icons/edit-icon.svg"
        />
        <img
          alt="focus icon"
          class="icon"
          src="/icons/focus-icon.svg"
        />
        <img
          alt="merge icon"
          class="icon"
          src="/icons/merge-icon.svg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
